### PR TITLE
Auth middleware

### PR DESCRIPTION
**Pull request by dee-me-tree-or-love**
**What**
Implemented the JWT middleware for authorization purposes

**Changes**
Started using the `express-jwt` powered _authorization required_ middleware
Created the _username match_ middleware
Added a very generic _error handling_ middleware

**Points**
- [x] Drafted out the middleware
- [x] Implemented the middleware
- [x] Added the middleware to the routes _check needed_
- [x] Tested middleware correctness _check needed_

**Help**
Would appreciate some help in preparing tests for the middleware to ensure it works as intended and also a fellow programmer's view on the code for the bug detecting purposes :upside_down_face:

### CODE REVIEW

File: src/controllers/users/index.js [hunk 1]
```diff
@@ -1,14 +1,23 @@
...
-router.put('/find', userHandler.getUserByUsername);
-router.post('/', userHandler.createUser);
-router.put('/', userHandler.updateUser);
-router.put('/delete', userHandler.deleteUserByUsername);
+router
+  .get('/', userHandler.getUsers);
+router
+  .put('/find', userHandler.getUserByUsername);
+router
+  .post('/', userHandler.createUser);
```
Comment (dee-me-tree-or-love) [inline src/controllers/users/index.js:14]: What kind of authorization should be checked for on this action? 
Internal client access token or something like that maybe? I am really not sure...

File: src/exceptions/UserServiceError.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+module.exports = class UserServiceException extends Error {
```
Comment (dee-me-tree-or-love) [inline src/exceptions/UserServiceError.js:1]: Yeah, it's a generic exception, but with a custom name - kinda might come in handy at some point

File: src/middleware/ErrorHandling.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * Handle an error that can occur during the route handling
+ * @param {Object} err
+ * @param {Object} req
+ * @param {Object} res
+ * @param {Object} next
+ * @return {Object}
+ */
+const handleError = (err, req, res, next) => {
+  console.error({
```
Comment (dee-me-tree-or-love) [inline src/middleware/ErrorHandling.js:12]: @dpapukchiev you had something like `req.log.error(...)` here - for me it gave an error - I didn't really look into what was the issue, but maybe you know on top of your head?

File: src/middleware/UsernameMatchRequired.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+/**
+ * Validates if the username resolved from the authorization
+ * token matches that from the body
+ * @param {*} req
+ * @param {*} res
+ * @param {Function} next
+ * @throws {Error}
+ */
+const userNameMatch = (req, res, next) => {
+  if (req.user.username !== req.body.username) {
```
Comment (dee-me-tree-or-love) [inline src/middleware/UsernameMatchRequired.js:12]: I mean that's it, right? :thinking:

**Comment by dee-me-tree-or-love** [comment]
- [x] Change the update point to create new user document if the username is yet unknown 
- [x] Remove the username from the rqeuired payload parameters and extract it from the access token instead
- [x] Coordinate with @dmarkov00 the access token handling idea _(cookie + local storage compromise)_

**Comment by dee-me-tree-or-love** [comment]
So now should be alright - the create endpoint is protected waiting for some kind of a token at least - we can figure the logic later... 
The update endpoint creates new users if `username` from access token is freshly met. 
Now should be alright.

**Review by dpapukchiev — DISMISSED** [review]

**Comment by dee-me-tree-or-love** [comment]
Damn, I might break everything with these cahnges, buuuuuut, YOLO - I am merging 👍

**Comment by dee-me-tree-or-love** [comment]
The changes broke everything - hence the revert
I am not fixing it

**Pull request merged**